sbor-tests/tests/enum.rs

```diff
@@ -0,0 +1,95 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sbor::rust::prelude::*;
+use sbor::*;
+
+const VARIANT_1: u8 = 4;
+const VARIANT_2: u8 = 5;
+
+#[derive(Sbor, PartialEq, Eq, Debug)]
+pub enum Abc {
+    #[sbor(id(VARIANT_1))]
+    Variant1,
+    #[sbor(id(VARIANT_2))]
+    Variant2,
+}
+
+const CONST_32: u8 = 32;
+const CONST_55: u8 = 55;
+
+#[derive(Sbor, PartialEq, Eq, Debug)]
+#[repr(u8)]
+pub enum Mixed {
+    #[sbor(id = 5)]
+    A,
+    #[sbor(id(7))]
+    B,
+    #[sbor(id("8"))]
```

Comment:
I'm a bit worried that we support all three syntaxes. 

Possible to limit to a single syntax with value being "any expression that evaluates into u8"?

Reply:
Hah I actually pre-empted this feedback with my latest comment and removed `id` before reading your message :) - we only support `discriminator` now.

This gives us then two core options (repr and `#[sbor(discriminator)]`).

I think if the user specifies a `repr` then it makes sense to use that for the discriminators by default (because I imagine at a low level the compiler can optimise away lookups/branches).

But forcing use of repr feels wrong (as it might affect performance in other ways) - so I wanted to provide an alternative mechanism, hence the sbor attribute.

Does that sound good to you?

Reply:
I don't think we should use `repr`, because users may `repr` to `non-u8` for other purposes. I feel we should be very SBOR-specific, even if it means users may have to annotate twice.

But I will leave the decision to you :)

Reply:
I take your point!

I've tweaked it so that we only use the `repr` values if we specify `#[sbor(use_repr_discriminators)]` - this way it won't happen by accident. And they can still be overridden with explicit annotations.